File: frontend/src/objects/rating/response/RatingContestListItem.ts
import type { ContestSlug } from '@/objects/contest/ContestSlug'
import type { ContestTitle } from '@/objects/contest/ContestTitle'
import type { UserIdentity } from '@/objects/user/UserIdentity'

/** Rating 管理中的比赛条目；包含比赛时间、参数、参与人数和追加人。 */
export type RatingContestListItem = {
  position: number
  contestSlug: ContestSlug
  contestTitle: ContestTitle
  contestStartAt: string
  contestEndAt: string
  m: number
  participantCount: number
  overlapWarning: boolean
  appendedBy: UserIdentity | null
  appendedAt: string
}

File: backend/src/main/scala/domains/rating/objects/response/RatingContestListItem.scala
package domains.rating.objects.response

import domains.contest.objects.{ContestSlug, ContestTitle}
import domains.user.objects.UserIdentity
import io.circe.{Decoder, Encoder}
import io.circe.generic.semiauto.{deriveDecoder, deriveEncoder}

import java.time.Instant
import scala.util.Try

/** 评分管理页中的已追加比赛条目，包含序列位置、快照元数据和追加人。 */
final case class RatingContestListItem(
  position: Int,
  contestSlug: ContestSlug,
  contestTitle: ContestTitle,
  contestStartAt: Instant,
  contestEndAt: Instant,
  m: Int,
  participantCount: Int,
  overlapWarning: Boolean,
  appendedBy: Option[UserIdentity],
  appendedAt: Instant
)

/** 提供评分比赛管理条目 JSON codec，并显式处理 Instant 字符串格式。 */
object RatingContestListItem:
  private given Encoder[Instant] = Encoder.encodeString.contramap(_.toString)
  private given Decoder[Instant] = Decoder.decodeString.emap(value => Try(Instant.parse(value)).toEither.left.map(_.getMessage))

  given Encoder[RatingContestListItem] = deriveEncoder[RatingContestListItem]
  given Decoder[RatingContestListItem] = deriveDecoder[RatingContestListItem]
